kiwi/lib/search.py

```diff
@@ -0,0 +1,438 @@
+    warmup_steps = trial.suggest_uniform(
+        'warmup_steps',
+        config.options.warmup_steps.lower,
+        config.options.warmup_steps.upper,
+    )
+    freeze_epochs = trial.suggest_uniform(
+        'freeze_epochs',
+        config.options.freeze_epochs.lower,
+        config.options.freeze_epochs.upper,
+    )
+
+    kiwi_config['system']['optimizer']['learning_rate'] = learning_rate
+    kiwi_config['system']['model']['decoder']['dropout'] = dropout
+    kiwi_config['system']['model']['outputs']['dropout'] = dropout
+    kiwi_config['system']['optimizer']['warmup_steps'] = warmup_steps
+    kiwi_config['system']['model']['encoder']['freeze_for_number_of_steps'] = int(
+        updates_per_epochs * freeze_epochs
+    )
+
+    # Iterable printables
+    names = ['learning_rate', 'dropout', 'warmup_steps', 'freeze_epochs']
+    values = [learning_rate, dropout, warmup_steps, freeze_epochs]
+
+    if config.options.hidden_size is not None:
+        hidden_size = trial.suggest_categorical(
+            'hidden_size', config.options.hidden_size
+        )
+        kiwi_config['system']['model']['encoder']['hidden_size'] = hidden_size
+        kiwi_config['system']['model']['decoder']['hidden_size'] = hidden_size
+        names.append('hidden_size')
+        values.append(hidden_size)
+
+    if config.options.bottleneck_size is not None:
+        bottleneck_size = trial.suggest_categorical(
+            'bottleneck_size', config.options.bottleneck_size
+        )
+        kiwi_config['system']['model']['decoder']['bottleneck_size'] = bottleneck_size
+        names.append('bottleneck_size')
+        values.append(bottleneck_size)
+
+    if config.options.search_mlp:
+        use_mlp = trial.suggest_categorical('mlp', [True, False])
+        kiwi_config['system']['model']['encoder']['use_mlp'] = use_mlp
+        names.append('use_mlp')
+        values.append(use_mlp)
+
+    # Search word_level and sentence_level and their combinations
+    if config.options.search_hter:
+        assert kiwi_config['data']['train']['output']['sentence_scores'] is not None
+        assert kiwi_config['data']['valid']['output']['sentence_scores'] is not None
+        hter = trial.suggest_categorical('hter', [True, False])
+        kiwi_config['system']['model']['outputs']['sentence_level']['hter'] = hter
+        names.append('hter')
+        values.append(hter)
+
+    if config.options.search_word_level:
+        assert kiwi_config['data']['train']['output']['target_tags'] is not None
+        assert kiwi_config['data']['valid']['output']['target_tags'] is not None
+        word_level = trial.suggest_categorical('word_level', [True, False])
+        kiwi_config['system']['model']['outputs']['word_level']['target'] = word_level
+        kiwi_config['system']['model']['outputs']['word_level']['gaps'] = word_level
+        names.append('word_level')
+        values.append(word_level)
+
+    if config.options.search_hter and config.options.search_word_level:
```

Comment:
I removed the first block, because it actually made no sense: search_hter=true and search_word_level=true should be allowed at the same time. I added a validator to catch this.